AP_HAL: fixed get_size() on ObjectBuffer to be consistent with set_si…

tridge opened the pull request:
Our ring buffers can hold at most N-1 elements if we allocate N elements worth of memory. The set_size() API reflects this, and allocates room for N+1 elements. If you then query the size with get_size() then you should get back the same number, not N+1. Otherwise you may think there is room for your data when there isn't

peterbarker commented:
This looks right, but space() also needs fixing.

https://github.com/ArduPilot/ardupilot/compare/master...peterbarker:pr/ringbuffer-tests?expand=1 contains a commit which adds unit tests for the ringbuffer class.

Rudimentary unit test suite failing on master here: https://github.com/ArduPilot/ardupilot/compare/master...peterbarker:pr/ringbuffer-tests?expand=1

rmackay9 commented:
Because i'm moaning at other developers I'll moan here too that I think we should always have a description.  for example, why is this important?

tridge commented:
> This looks right, but space() also needs fixing.

no, space() is already correct, as the integer division rounds down. If you have a buffer of 32 bit integers and use set_size(8) then it allocates 32 bytes. With this PR get_size() returns 8. The space() call also returns 8, as space() on the underlying byte buffer returns 35, which when divided by 4 gives 8.
The test_ringbuffer.cpp code is wrong. I've fixed the test and added it to this PR

tridge commented:
> Because i'm moaning at other developers I'll moan here too that I think we should always have a description. for example, why is this important?

yep, fixed

peterbarker commented:
Merged, thanks!
